components/carousel/utils/useCarouselEvents.ts

```diff
@@ -0,0 +1,151 @@
+import type { CarouselNFT } from '@/components/base/types'
+import type { NFTWithMetadata } from '@/composables/useNft'
+import { isBeta, isProduction } from '@/utils/chain'
+import { formatNFT } from '@/utils/carousel'
+
+import latestEvents from '@/queries/subsquid/general/latestEvents.graphql'
+import latestEventsRmrkv2 from '@/queries/subsquid/ksm/latestEvents.graphql'
+
+interface Types {
+  type: 'latestSales' | 'newestList'
+}
+
+const limit = isProduction ? 15 : 8
+
+const nftEventVariables = {
+  latestSales: { interaction_eq: 'BUY' },
+  newestList: { interaction_eq: 'LIST' },
+}
+
+const disableChainsOnBeta = ['snek']
+
+const fetchLatestEvents = (chain, type, where = {}) => {
+  const { $apollo } = useNuxtApp()
+  const query = chain === 'ksm' ? latestEventsRmrkv2 : latestEvents
+
+  return $apollo.query({
+    query,
+    client: chain,
+    variables: {
+      limit: limit * 2, // buffer the limit
+      orderBy: 'timestamp_DESC',
+      where: {
+        ...nftEventVariables[type],
+        ...where,
+      },
+    },
+  })
+}
+
+const useChainEvents = (chain, type) => {
```

Comment:
<img src="https://media2.giphy.com/media/2sk6tdx6Nv4BQBwlsP/giphy.gif"/>

Reply:
anyway, is there any plugin for codeclimate to detect vue composition API?

imo, codeclimate is not able to scan composable functions properly. in pure function, I think it still makes sense, but in composable function quite tricky to solve it

because, in composable the structure is something like this
```js
const useComposableName = () => {
  // we put state here or call another composable
  const state = ref('')

  // and then, we put methods here that need to access the state above
  // sometimes, we need to put several methods to access the same state. for example on `useMetaTransaction.ts`
  // in Vue Options API, the method is wrapped under `methods` property. I think codeclimate will pass that
  // but, codeclimate sum up the complexity from `useComposableName` instead of separate that from each method
  const setName = (name: string) => {
    state.value = name
  }
}
```
![image](https://github.com/kodadot/nft-gallery/assets/734428/83543fb1-1216-4f5b-9d61-e61dd6309088)

Reply:
for comparison, this is the code example above on Vue Options API. codeclimate won't sum up the complexity because Options API uses `Object` (`export default {}`). meanwhile, composable is using function `function useComposableName`. so codeclimate assumed that it was a pure function. Imo, codeclimate should treat OptionsAPI and CompositionAPI the same

```
export default {
  data() {
    return {
      state: '',
    }
  },
  methods: {
    setName(name: string) {
      this.state = name
    },
  },
}
```

Reply:
anyway, I already put my opinion in here related codeclimate on composable function https://github.com/kodadot/nft-gallery/pull/5765#issuecomment-1532322967